File: observalog-go/middleware/kafka_consumer.go
package middleware

import (
	"context"
	"sync/atomic"

	log "github.com/darshanredkar11/observalog-go"
)

// Message represents a Kafka message with headers.
type Message interface {
	Headers() map[string]string
}

// KafkaMiddleware wraps a Kafka message handler to inject trace context.
// The handler signature is handler(ctx context.Context, msg Message) error.
func KafkaMiddleware(handler func(context.Context, Message) error) func(context.Context, Message) error {
	return func(ctx context.Context, msg Message) error {
		headers := msg.Headers()

		// Extract or propagate trace_id.
		if traceID, ok := headers["X-Trace-Id"]; ok {
			ctx = log.WithTraceID(ctx, traceID)
		}

		// Extract parent span_id.
		if parentSpan, ok := headers["X-Span-Id"]; ok {
			ctx = log.WithParentSpan(ctx, parentSpan)
		}

		// Extract user_id.
		if userID, ok := headers["X-User-Id"]; ok {
			ctx = log.WithUserID(ctx, userID)
		}

		// Extract journey_stage.
		if journeyStage, ok := headers["X-Journey-Stage"]; ok {
			ctx = log.WithJourneyStage(ctx, journeyStage)
		}

		// Create fresh seq counter for this message.
		ctx = log.WithSeq(ctx, new(atomic.Uint32))

		return handler(ctx, msg)
	}
}

File: observalog-node/src/middleware/kafka-consumer.ts
import { runWithContextAsync } from '../context';

export interface KafkaMessage {
    headers?: Record<string, Buffer | string | undefined>;
}

// Wraps a KafkaJS message handler with trace context extracted from message headers.
//
// Usage:
//   consumer.run({
//     eachMessage: kafkaConsumerMiddleware(async ({ message }) => {
//         log.info('provider.message.received', 'Message received', { topic });
//     }),
//   });
export function kafkaConsumerMiddleware<T extends { message: KafkaMessage }>(
    handler: (payload: T) => Promise<void>,
): (payload: T) => Promise<void> {
    return (payload: T): Promise<void> => {
        const h = (name: string): string => {
            const v = payload.message.headers?.[name];
            if (!v) return '';
            return Buffer.isBuffer(v) ? v.toString('utf8') : String(v);
        };

        const ctx = {
            traceId:      h('X-Trace-Id'),
            spanId:       h('X-Span-Id'),
            parentSpan:   h('X-Parent-Span-Id'),
            userId:       h('X-User-Id'),
            journeyStage: h('X-Journey-Stage'),
            seq:          { value: 0 },
        };

        return runWithContextAsync(ctx, () => handler(payload));
    };
}
